Wikipedia/Code/ArticleSurveyTimerController.swift

```diff
@@ -0,0 +1,125 @@
+import Foundation
+
+/// Manages the timer used to display survey announcements on articles
+final class ArticleSurveyTimerController {
+
+    // MARK: - Public Properties
+
+    var timerFireBlock: ((SurveyAnnouncementsController.SurveyAnnouncementResult) -> Void)?
+
+    // MARK: - Private Properties
+
+    private let articleURL: URL
+    private let surveyController: SurveyAnnouncementsController
+
+    private var timer: Timer?
+    private var timeIntervalRemainingWhenBackgrounded: TimeInterval?
+    private var shouldPauseOnBackground = false
+    private var shouldResumeOnForeground: Bool { return shouldPauseOnBackground }
+
+    // MARK: - Computed Properties
+
+    private var surveyAnnouncementResult: SurveyAnnouncementsController.SurveyAnnouncementResult? {
+        get {
+            guard let articleTitle = articleURL.wmf_title?.denormalizedPageTitle, let siteURL = articleURL.wmf_site else {
+                    return nil
+            }
+
+            return SurveyAnnouncementsController.shared.activeSurveyAnnouncementResultForTitle(articleTitle, siteURL: siteURL)
+        }
+    }
+
+    // MARK: - Lifecycle
+
+    init(articleURL: URL, surveyController: SurveyAnnouncementsController) {
+        self.articleURL = articleURL
+        self.surveyController = surveyController
+    }
+
+    // MARK: - Public
+
+    func articleContentDidLoad() {
```

Comment:
Not a bug, just extra code - 

We don't check that `surveyAnnouncementResult` is non-nil here and we don't set `shouldPauseOnBackground` as `true` - as we do in `viewWillAppear` below. This is ultimately ok, because these things happen in `startTimer`. But given that it happens in `startTimer`, do we need these two things in `viewWillAppear`? In `viewWillAppear`, these two things happen twice - once in `viewWillAppear`, and once in `startTimer`.

Reply:
Apologies for not addressing this one earlier. Great point about potentially extraneous code. There are definitely some defensive programming strategies here, which I think are in large part helpful given the global mutability within `surveyAnnouncementResult` and async nature of web requests. The `viewWillAppear` and `viewWillDisappear` assignments in particular work together to mitigate issues surrounding 5, 6, and 7 from https://github.com/wikimedia/wikipedia-ios/pull/3554#issuecomment-624346601. Otherwise you run into unexpected timer behavior navigating the article stack and backgrounding/foregrounding the app.

I'd be happy to continue looking into optimizations here, but I wouldn't feel comfortable with us committing/merging changes to the timer start/stop/pause/resume logic unless we all had an opportunity to do another full round of tests with the updated logic. For that reason, if you're comfortable with it, may be worth merging and getting functional in-use feedback from the beta users.

Reply:
Per elsewhere, I'm definitely comfortable merging it. Will do that shortly. 

That makes sense about being defensive around all this. We may also (at some point in the future) want to add the extra defensive check to `articleContentDidLoad` - that's the only one that doesn't have it.